codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java

```diff
@@ -634,15 +669,17 @@ public void appendToHeaderList(CharSequence name, CharSequence value) {
                 return;
             }
 
-            if (validate) {
-                try {
-                    previousType = validate(streamId, name, previousType, headers, value);
-                } catch (Http2Exception ex) {
-                    validationException = ex;
-                    return;
+            try {
+                if (validateHeaders) {
+                    previousType = validateHeader(streamId, name, previousType, headers, value);
+                }
+                if (validateHeaderValues) {
+                    validateValidHeaderValue(streamId, name, value);
```

Comment:
Not sure if it matters but should we set `previousType` to the old value again as `validateHeaders` might throw after we have update it.

Reply:
Moving value validation before the regular validation would solve that.